Unicode/ustring.c

```diff
@@ -301,33 +301,46 @@ return( NULL );
 
 unichar_t *u_copyn(const unichar_t *pt, long n) {
     unichar_t *res;
+
+    if ( pt && ++n>0 ) {
 #ifdef MEMORY_MASK
-    if ( n*sizeof(unichar_t)>=MEMORY_MASK )
-	n = MEMORY_MASK/sizeof(unichar_t)-1;
+	if ( n>MEMORY_MASK/sizeof(unichar_t) )
+	    n = MEMORY_MASK/sizeof(unichar_t);
 #endif
-    res = (unichar_t *) malloc((n+1)*sizeof(unichar_t));
-    memcpy(res,pt,n*sizeof(unichar_t));
-    res[n]='\0';
-return(res);
+	if ( (res=(unichar_t *)(malloc(n*sizeof(unichar_t)))) ) {
+	    if ( (--n) ) memcpy(res,pt,n*sizeof(unichar_t));
+	    res[n] = 0;
+	    return( res );
+	}
+    }
+    return( NULL );
 }
 
 unichar_t *u_copynallocm(const unichar_t *pt, long n, long m) {
+/* Alloc space for m unichar chars and copy n unichar chars */
     unichar_t *res;
+
+    if ( pt && ++n>0 && ++m>0 && m>=n ) {
 #ifdef MEMORY_MASK
-    if ( n*sizeof(unichar_t)>=MEMORY_MASK )
-	n = MEMORY_MASK/sizeof(unichar_t)-1;
+	if ( n>MEMORY_MASK/sizeof(unichar_t) )
+	    n = MEMORY_MASK/sizeof(unichar_t);
 #endif
-    res = malloc((m+1)*sizeof(unichar_t));
-    memcpy(res,pt,n*sizeof(unichar_t));
-    res[n]='\0';
-return(res);
+	if ( (res=(unichar_t *)(calloc(m,sizeof(unichar_t)))) ) {
+	    if ( (--n) ) memcpy(res,pt,n*sizeof(unichar_t));
+	    return( res );
+	}
+    }
+    return( NULL );
 }
 
 unichar_t *u_copy(const unichar_t *pt) {
-    if(pt)
-return u_copyn(pt,u_strlen(pt));
+/* Copy unichar_t string and if error, return( "" ). */
+    long n;
+
+    if ( pt && (n=u_strlen(pt))>0 )
```

Comment:
Why, where did you actually see this be an issue; `u_copyn` with `u_strlen` returning 0 would have been valid in the old case anyway

Reply:
testing for failing
if ( pt=NULL || n<0 || n+1<0 )
testing for pass
if ( pt!=NULL && n>=0 && n+1>0 )
or testing for pass (n+1)
if ( pt!=NULL && n>0 && n>0 )

Reply:
@JoesCat, I still don't understand. What value of pt and pointed contents would behave differently between these two cases? The only one I can think of is one in which n gets overflowed.

Reply:
for n==0 there wasn't a need to go through the exercise of testing variables, malloc and insert 0 as there's already was a calloc(1,sizeof(unichar_t)) below that did the exact same thing.

Reply:
Yes, but you added the calloc call, right? Was the ambiguity of sometimes returning a static string and other times a dynamic string the thing you were trying to fix?

Reply:
Fixing the static string was definitely the main interest.
When you look at calls used everywhere, there is a good potential to cleanup and free memory instead of losing and leaking memory everywhere.
There is a calloc added internally.
Looking at all the feedback here, another commit is needed, the git comment would look pretty ugly on a squash commit, so probably better to create a fresh commit.

Reply:
@JoesCat, I still don't understand some of what you've done. Let me finish up comments tomorrow, and then we can finalize changes. The changeset is of sufficiently singular purpose to go all into one giant commit, I think, as long as the functionality is appropriately commented.

Reply:
I don't think that this actually saves significant execution time in the zero-length case, which is rare, and it increases execution time in the other cases (two function calls).

Reply:
two function calls? where?